src/nupic/regions/BacktrackingTMRegion.cpp

```diff
@@ -0,0 +1,1191 @@
+    for (size_t i = 0; i < args_.numberOfCols * args_.cellsPerColumn; i++) {
+      ptr[i] = output[i];
+    }
+  }
+
+  if (args_.topDownMode) {
+    // Top - down compute
+    Real *tdout = tm_->topDownCompute();
+    getOutput("topDownOut")->getData() = Array(NTA_BasicType_Real32, 
+                                               tdout, 
+                                               args_.numberOfCols);
+  }
+
+  // Set output for use with anomaly classification region if in anomalyMode
+  if (args_.anomalyMode) {
+    Byte *lrn = tm_->getLearnActiveStateT();
+    Size size = args_.numberOfCols * args_.cellsPerColumn;
+    getOutput("lrnActiveStateT")->getData() = Array(NTA_BasicType_Byte, lrn, size);
+
+    auto activeColumns = VectorHelpers::binaryToSparse(bottomUpIn.asVector<Real32>());
+    Real32 anomalyScore = algorithms::anomaly::computeRawAnomalyScore(
+        activeColumns, prevPredictedColumns_);
+    getOutput("anomalyScore")->getData() = Array(NTA_BasicType_Real32, &anomalyScore, 1);
+  }
+
+  if (args_.computePredictedActiveCellIndices) {
+    Output *activeCells = getOutput("activeCells");
+    Output *predictedActiveCells = getOutput("predictedActiveCells");
+    Byte *activeState = tm_->getActiveState();
+    Size nCells = args_.numberOfCols * args_.cellsPerColumn;
+    NTA_ASSERT(activeCells != nullptr);
+    NTA_ASSERT(predictedActiveCells != nullptr);
+    NTA_ASSERT(args_.outputWidth == nCells);
+    NTA_ASSERT(args_.outputWidth == activeCells->getData().getCount());
+    NTA_ASSERT(args_.outputWidth == predictedActiveCells->getData().getCount());
+
+    Real32 *activeCellsPtr = (Real32 *)activeCells->getData().getBuffer();
+    Real32 *predictedActiveCellsPtr =  (Real32 *)predictedActiveCells->getData().getBuffer();
+    for (size_t idx = 0; idx < nCells; idx++) {
+      activeCellsPtr[idx] = (activeState[idx]) ? 1.0f : 0.0f;
+      predictedActiveCellsPtr[idx] = (prevPredictedState_[idx] && activeState[idx]) ? 1.0f : 0.0f;
+    }
+  }
+}
+
+std::string BacktrackingTMRegion::executeCommand(const std::vector<std::string> &args,
+                                     Int64 index) {
+  // The TM does not execute any Commands.
+  return "";
+}
+
+// This is the per-node output size. This determines how big the output
+// buffers should be allocated to during Region::initialization(). NOTE: Some
+// outputs are optional, return 0 if not used.
+size_t BacktrackingTMRegion::getNodeOutputElementCount(const std::string &outputName) const {
+  if (outputName == "bottomUpOut")
+    return (args_.orColumnOutputs)? args_.numberOfCols: args_.outputWidth;
+  if (outputName == "topDownOut")
+    return args_.numberOfCols;
+  if (outputName == "lrnActiveStateT")
+    return args_.outputWidth;
+  if (outputName == "activeCells")
+    return args_.outputWidth;
+  if (outputName == "predictedActiveCells")
+    return args_.outputWidth;
+  return 0; // an optional output that we don't use.
```

Comment:
> NTA_THROW << "unexpected"

No, this is OK.  The framework is just checking all of the outputs declared in the spec.  The region implementation has the right to decide if the parameters are such that it will or will not be able to use this output.